compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp

```diff
@@ -1001,6 +1001,82 @@ class MoveAllDependenciesIntoOp : public OpRewritePattern<OpTy> {
   }
 };
 
+/// Utility to convert function arguments with memref element types' bitwidth
+/// either 8 or 16 to i32. This basically involves linearizing an N-d memref to
+/// 1-d i32 memref and adjusting the offset/size/stride metadata.
+static LogicalResult CastFunctionArgs(func::FuncOp funcOp,
+                                      PatternRewriter &rewriter) {
+  // only run on npu control functions
+  bool hasNpuOps = false;
+  funcOp.walk([&](AIEX::NpuDmaMemcpyNdOp dma) { hasNpuOps = true; });
+  if (!hasNpuOps) return failure();
+
+  // Cast all the function args to i32 types.
+  // This is in support of npu.dma_memcpy_nd which only allow 32bit integer
+  // types.
+  mlir::FunctionType funcType = funcOp.getFunctionType();
+  SmallVector<Type> argTypes(funcType.getInputs());
+  for (int i = 0, e = argTypes.size(); i < e; i++) {
+    auto memrefTy = dyn_cast<MemRefType>(argTypes[i]);
+    if (!memrefTy) continue;
+
+    unsigned int bitwidth = memrefTy.getElementTypeBitWidth();
+    if (bitwidth != 16 && bitwidth != 8) continue;
+    unsigned upcastingDivisor = 32 / bitwidth;
+
+    unsigned int div = 32 / bitwidth;
+    unsigned int numElements = memrefTy.getNumElements() / div;
+    SmallVector<int64_t> shape{numElements};
+    MemRefType newMemrefTy =
+        MemRefType::get(shape, rewriter.getIntegerType(32));
+    argTypes[i] = newMemrefTy;
+    auto &entry = funcOp.front();
+    entry.insertArgument(i, newMemrefTy, rewriter.getUnknownLoc());
+    rewriter.setInsertionPointToStart(&entry);
+    // With memref shape collapsed to 1d, the multi-dimensional offset also
+    // needs to be collapsed.
+    SmallVector<Operation *> users;
+    for (Operation *user : entry.getArgument(i + 1).getUsers()) {
+      if (auto dmaUser = dyn_cast<AIEX::NpuDmaMemcpyNdOp>(user)) {
+        int oneDOffset = 0;
+        for (int j = 0, n = dmaUser.getMixedOffsets().size(); j < n; j++)
+          oneDOffset += *getConstantIntValue(dmaUser.getMixedOffsets()[j]) *
+                        *getConstantIntValue(dmaUser.getMixedStrides()[j]);
+        rewriter.setInsertionPoint(dmaUser);
+        SmallVector<int64_t, 4> newStaticOffsets = {0, 0, 0, oneDOffset};
```

Comment:
actually I dont follow why we are restricted to make 4 dims always, we can just keep the same dims as the op already has?

Reply:
This is majorly because of restrictions in the lower stack. Here is [aiex.npu.dma_memcpy_nd's verify()](https://github.com/Xilinx/mlir-aie/blob/main/lib/Dialect/AIEX/IR/AIEXDialect.cpp#L187-L193) - it makes certain assumptions on size/stride and ideally we would want all the metadata to be of same shape.